Split out Tamanu API client

passcod opened the pull request:
### Changes

Create new package `@tamanu/api-client` which contains a client for the Tamanu (internal) API, for use within other Tamanu packages. This is the client defined in desktop, though at a latter point it might be interesting to join to that the client defined in facility/lan for sync.

~~Includes a CLI API client:~~

![image](https://github.com/beyondessential/tamanu/assets/155787/cb9c6b79-5e98-404e-bd13-564d4021ef3c)


### Checklist

- [x] Code is finished
- [x] CI passes (dependent on upstream PR)

passcod committed:
Make package for Tamanu API Client

passcod committed:
Move things around

passcod committed:
Attempt to fix eslint errors

passcod committed:
Reformat

passcod committed:
Add api-client to dockerfile

passcod committed:
Test

passcod committed:
Adjust lint configs

passcod committed:
Emit types in the right place

passcod committed:
Well that was a nightmare and a half

passcod committed:
Whooops

passcod committed:
More mass replace constants imports

passcod committed:
Run yarn

passcod committed:
Handle 204s in Tamanu client

passcod committed:
Style fix

passcod committed:
Add CLI api client

passcod committed:
API: Split user load from login, to allow full hydration when loading a token

passcod committed:
API: Restore misplaced isErrorUnknowns

passcod committed:
Revert .new-package

passcod committed:
Lint constants

passcod committed:
Revert more files from merge

passcod committed:
Actually throw response

passcod committed:
Fix package.json

passcod committed:
Run yarn

mcccclean reviewed (changes requested):
This is really neat! Just one change request but will yield in the face of much pushback on it.

passcod committed:
Merge branch 'main' into fix/tamanu-api-client

passcod committed:
Handle 404s in api-client package

passcod committed:
Merge branch 'main' into fix/tamanu-api-client

passcod committed:
Disable incremental TS

Actually a huge pain to work with, given it doesn't regenerate .d.ts if they're missing as long as the .tsbuildconfig is "up to date" with the source. Why oh why must typescript insist on implementing things the wrong way.

passcod committed:
Remove CLI

passcod committed:
I think that's right
